### ISSUE TITLE
getting peerId (or peerInfo) instance in transport

### ISSUE THREAD

**Issue by pgte** [libp2p/js-libp2p#222]
Currently, the libp2p transport and discovery modules config accepts instances or classes. Inside the transport constructor call I need access to the `peerInfo` instance. Could this be done inside [the constructor call](https://github.com/libp2p/js-libp2p/blob/8533936ffa9c56436dfdb6825cd83000b46c853c/src/index.js#L133) ?

**Comment by jacobheun** [comment]
The transport interface currently doesn't take peerInfo as an argument for the constructor. Are you using a custom transport? I'm curious what the use case is to see if updating the interface would be useful. 

Since the peerInfo needs to be known for libp2p instance creation, you can currently work around this by passing that to a new transport instance and then using the transport instance during libp2p configuration. Does this approach create problems for you?

**Comment by pgte** [comment]
@jacobheun yes, I'm using a custom transport that wraps another transport, see [here](https://github.com/ipfs-shipyard/peer-star-app/blob/master/src/transport/ipfs.js#L26).

What workaround were you proposing?

**Comment by jacobheun** [comment]
Okay, so you're doing what I was talking about where you are initializing websocketstar, but it looks like that becomes seemingly impossible to do through the ipfs abstraction.

I feel like the right thing to do here is to have js-ipfs allow for custom libp2p generators to be added via configuration. It currently allows for providing a repo instance, but libp2p would need some of the info from initialization. So being able to provide a libp2p generator function instead of configuration feels like that could make things easier for developers to be able to customize libp2p to their needs.

This would allow you to do something like this:
```js
  const ipfs = new IPFS({
    libp2p: (_ipfs) => {
      return new Libp2p({
        peerInfo: _ipfs._peerInfo,
        peerBook: _ipfs._peerInfoBook,
        config: { /*custom config*/ }
      })
    }
  })
```

@alanshaw what do you think about this approach?

**Comment by pgte** [comment]
@jacobheun that would help if we the generator somehow gets a peerId like you showed.

**Comment by jacobheun** [comment]
I've got an open PR for this support, ipfs/js-ipfs#1470. I included a working [example of configuration](https://github.com/ipfs/js-ipfs/blob/6543854dbb2fac6ffe34b68fe01aa35a8ffa0ae7/examples/custom-libp2p/index.js) as well, that shows creating a transport with the ipfs nodes peerId

**Comment by jacobheun** [comment]
Support for this was released in js-ipfs@0.32.0
